File: 2022/09/9.clj
(ns aoc.2022.09.9
  (:require [aoc.common :refer [read-input]]
            [clojure.string :as str]))

(defn move-head [[hx hy] dir]
  (case dir
    "U" [hx (dec hy)]
    "D" [hx (inc hy)]
    "L" [(dec hx) hy]
    "R" [(inc hx) hy]))

(defn move-tail [[hx hy] [tx ty]]
  (let [dx (- hx tx) dy (- hy ty)
        adx (abs dx) ady (abs dy)
        nx (cond (= 2 adx) (/ 2 dx)
                 (= 2 ady) dx
                 :else 0)
        ny (cond (= 2 ady) (/ 2 dy)
                 (= 2 adx) dy
                 :else 0)]
    [(+ tx nx) (+ ty ny)]))

(defn move [rope [dir dist]]
  (if (zero? dist)
    rope
    (let [[head & tails] rope
          new-head (move-head (peek head) dir)
          new-tails (->> tails
                         (reduce (fn [[prev tails] tail]
                                   (let [nt (move-tail prev (peek tail))]
                                     [nt (conj tails (conj tail nt))]))
                                 [new-head []])
                         peek)]
      (move (apply conj [(conj head new-head)] new-tails)
            [dir (dec dist)]))))

(defn solve [moves rope]
  (->> moves
       (reduce move rope)
       peek
       set
       count))

(let [moves (->> (read-input)
                 (map #(str/split % #" "))
                 (map (fn [[d i]] [d (parse-long i)])))]
  
  (println "Part 1:" (solve moves [[[0 0]] [[0 0]]]))
  (println "Part 2:" (solve moves (repeat 10 [[0 0]]))))

File: 2022/09/9.py
moves = [[line.split()[0], int(line.split()[1])] for line in open("input.txt").read().splitlines()]

def move_head(hx, hy, dir):
  return {
    "U": (hx, hy-1),
    "D": (hx, hy+1),
    "L": (hx-1, hy),
    "R": (hx+1, hy)
  }[dir]

def move_tail(hx, hy, tx, ty):
  dx = hx - tx
  dy = hy - ty
  nx = dx//2 if abs(dx) == 2 else dx if abs(dy) == 2 else 0
  ny = dy//2 if abs(dy) == 2 else dy if abs(dx) == 2 else 0
  return (tx+nx, ty+ny)

def move(rope, dir, dist):
  head = rope[0]
  tails = rope[1:]
  for _ in range(dist):
    head.append(move_head(*head[-1], dir))
    prev = head[-1]
    for tail in tails:
      tail.append(move_tail(*prev, *tail[-1]))
      prev = tail[-1]
  return rope

rope = [[(0, 0)] for _ in range(10)]
for m in moves:
  rope = move(rope, *m)

print("Part 1:", len(set(rope[1])))
print("Part 2:", len(set(rope[-1])))